File: src/native/printer-plugin/registry/connection-registry.ts
import type { PrinterConnectionType } from '../definitions'
import type { PrinterConnection } from '../internal/types'

export class ConnectionRegistry {
  private readonly factories = new Map<PrinterConnectionType, () => PrinterConnection>()

  register(type: PrinterConnectionType, factory: () => PrinterConnection): void {
    this.factories.set(type, factory)
  }

  resolve(type: PrinterConnectionType): PrinterConnection {
    const factory = this.factories.get(type)
    if (!factory) {
      throw new Error(`Connection "${type}" chưa được đăng ký.`)
    }
    return factory()
  }
}

File: android/app/src/main/java/com/ndtcore/pos/printer/registry/ConnectionRegistry.kt
package com.ndtcore.pos.printer.registry

import com.ndtcore.pos.printer.connection.PrinterConnection

class ConnectionRegistry {
    private val factories = mutableMapOf<String, () -> PrinterConnection>()

    fun register(type: String, factory: () -> PrinterConnection) {
        factories[type] = factory
    }

    fun resolve(type: String): PrinterConnection {
        val factory = factories[type]
            ?: throw IllegalArgumentException("Connection \"$type\" chưa được đăng ký.")
        return factory()
    }
}
